### PR TITLE
Add before_each and after_each to spec dsl

### PR DESCRIPTION
**Pull request by ibash**
Was trying to use `before_each` / `after_each` in something I was writing and found that it was missing.
Pleasantly surprised to find out clean this code base is, this only took a few minutes to find and fix!

### CODE REVIEW

**Comment by jhass** [comment]
The hooks are not cleaned up after a describe block, they're global.

**Comment by ibash** [comment]
Thanks for pointing that out. What do you think about making `before_each` and `after_each` local to a context instead of global. Since the root context is already there, you could achieve the same global `before/after` functionality by calling them before any of your tests.

**Comment by waterlink** [comment]
Read the `spec` code more carefully. Currently it is running `it` and `describe` blocks upon encountering. Meaning that:

```crystal
describe "example" do
  it "says hello" do
    puts "hello"
  end

  before_each { puts "before" }
  after_each { puts "after" }
end
```

should output:

```
before
hello
after
```

but currently, `hello` will be printed even before `before_each` called. In that case, you need to always have `before_each` block at the beginning.

`after_each` does not have this problem though.

File: src/spec/context.cr [hunk 1]
```diff
@@ -143,6 +148,22 @@ module Spec
...
+    def self.after_each(&block)
+      @@contexts_stack.last.after_each << block
+    end
+
+    def self.run_before_each_hooks
+      @@contexts_stack.map(&.before_each).flatten().each(&.call)
+    end
+
+    def self.run_after_each_hooks
+      @@contexts_stack.map(&.after_each).flatten().each(&.call)
```
Comment (ibash) [inline src/spec/context.cr]: Should probably reverse this so you get: "before1" "before2" "after2" "after1"

**Comment by ibash** [comment]
@waterlink Yes I understand that. A better way to deal with it would probably to build the tree of contexts, hooks, and tests before executing any test.

I can understand if we don't want expose the hooks because we don't want code to become dependent on odd behavior.

For what it's worth I made the hooks local to a context in the updated diff.

**Comment by jhass** [comment]
I'm still not sure if it's worth it, the current architecture doesn't allow to share data (like instance variables) between the hooks and the examples, there's no automatic cleanup of such data etc. I feel like adding these would paint a false picture. Given these constraints, what real world usecases do you have?

**Comment by waterlink** [comment]
I have tried to build this (actually tried to do it `RSpec` style with creation of class per each context and example). It made compiler test suite compile ~20 times longer..

You can take a look here as well: https://github.com/waterlink/spec2.cr - this library aims to do exactly what you want (it is not finished though, in a prototype state now).

**Comment by ibash** [comment]
@jhass fair points, @waterlink interesting, thanks for the link.
Use case wise I actually ended up implementing my tests without a before hook. With respect to that I'm okay closing this out and letting these problems be solved by a third party testing library.

**Comment by jhass** [comment]
Thanks for your effort anyway!

**Pull request closed without merging**